cmd/zoekt-sourcegraph-indexserver/cleanup.go

```diff
@@ -201,24 +215,84 @@ func getShards(dir string) map[uint32][]shard {
 			continue
 		}
 
-		repos, _, err := zoekt.ReadMetadataPathAlive(path)
+		var repos []*zoekt.Repository
+		switch opt {
+		case all:
+			repos, _, err = zoekt.ReadMetadataPath(path)
+		case alive:
+			repos, _, err = zoekt.ReadMetadataPathAlive(path)
+		}
 		if err != nil {
 			debug.Printf("failed to read shard: %v", err)
 			continue
 		}
 
 		for _, repo := range repos {
 			shards[repo.ID] = append(shards[repo.ID], shard{
-				RepoID:   repo.ID,
-				RepoName: repo.Name,
-				Path:     path,
-				ModTime:  fi.ModTime(),
+				RepoID:        repo.ID,
+				RepoName:      repo.Name,
+				Path:          path,
+				ModTime:       fi.ModTime(),
+				RepoTombstone: repo.Tombstone,
 			})
 		}
 	}
 	return shards
 }
 
+// tombstoneDuplicates finds duplicate shards in indexDir and sets tombstones
+// to all of them but one. If at most one of the duplicates was without a tombstone,
+// the function is a no-op. Otherwise, the one that was modified the latest remains
+// alive.
+// The assumption is that duplicate shards exist only as parts of compound shards:
+// we haven't seen simple duplicate shards in the wild.
+// Duplicate shards should not occur at all but we rarely see them due to a bug somewhere (probably fixed).
+// This function is intended to wait for a subsequent call to vacuum that will delete the
+// tombstoned repos.
+func tombstoneDuplicates(indexDir string) {
+	index := getShards(indexDir, all)
+
+	for repoID, shardsAll := range index {
+		shards := make([]shard, 0, len(shardsAll))
+		for _, s := range shardsAll {
+			// One repoID corresponds to multiple shards if either
+			// a) it is a large repo split across multiple shards, or
+			// b) it is a repo that is duplicated in several compound shards.
+			// We are only interested in case b).
+			if strings.HasPrefix(filepath.Base(s.Path), "compound-") {
+				shards = append(shards, s)
+			}
+		}
+		if len(shards) <= 1 {
+			continue
+		}
+
+		log.Printf("found %d duplicate shards for repoID=%d. Tombstoning all but one", len(shards), repoID)
+
+		maybeAliveIdx := 0
+		for i, s := range shards {
+			var mustChange bool
+			if shards[maybeAliveIdx].RepoTombstone == s.RepoTombstone {
+				mustChange = shards[maybeAliveIdx].ModTime.Before(s.ModTime)
+			} else {
+				mustChange = shards[maybeAliveIdx].RepoTombstone
+			}
+			if mustChange {
+				maybeAliveIdx = i
+			}
+		}
+		if maybeAliveIdx != 0 {
+			shards[0], shards[maybeAliveIdx] = shards[maybeAliveIdx], shards[0]
+		}
```

Comment:
It took me some time to verify that this works. Couldn't we just search for the latest timestamp, then swap, then untombsone index 0, then tombstome everything else?

```suggestion
	        latest := 0
		for i, s := range shards {
			if s.ModTime.After(shards[latest].ModTime) {
			            latest = i
			}
		}
		if latest != 0 {
			shards[0], shards[latest] = shards[latest], shards[0]
		}
```

Reply:
1. I wanted this function to be idempotent and in your suggestion we may change the only alive shard if a repo is updated between two calls. I don't think it is a problem.
2. The other reason is that the API in cleanup.go is better suited for setting tombstones than for unsetting them.

I've implemented your suggestion.